File: sdks/middleware/AuthServiceAuthorization.java
package com.authservice.middleware;

import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import javax.servlet.Filter;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletResponse;

public final class AuthServiceAuthorization {
    private AuthServiceAuthorization() {}

    public static String permissionFor(String resource, String action) {
        String normalizedResource = resource == null ? "" : resource.trim().toLowerCase();
        String normalizedAction = action == null ? "" : action.trim().toLowerCase();
        if (normalizedResource.isEmpty() || normalizedAction.isEmpty()) {
            return "";
        }
        return normalizedResource + ":" + normalizedAction;
    }

    @SuppressWarnings("unchecked")
    public static boolean isAuthorized(Map<String, Object> claims, String resource, String action) {
        if (claims == null) {
            return false;
        }
        if ("owner".equals(claims.get("org_role"))) {
            return true;
        }
        Object permissions = claims.get("org_permissions");
        return permissions instanceof List && ((List<String>) permissions).contains(permissionFor(resource, action));
    }

    public static Filter servletFilter(String resource, String action, Function<ServletRequest, Map<String, Object>> claimsGetter) {
        return new Filter() {
            @Override
            public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
                if (!isAuthorized(claimsGetter.apply(request), resource, action)) {
                    HttpServletResponse http = (HttpServletResponse) response;
                    http.setStatus(403);
                    http.setContentType("application/json");
                    http.getWriter().write("{\"error\":\"forbidden\"}");
                    return;
                }
                chain.doFilter(request, response);
            }
        };
    }
}

File: sdks/middleware/AuthServiceAuthorization.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Claims;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;

namespace AuthService.Middleware
{
    public static class AuthServiceAuthorization
    {
        public static string PermissionFor(string resource, string action)
        {
            resource = (resource ?? "").Trim().ToLowerInvariant();
            action = (action ?? "").Trim().ToLowerInvariant();
            return resource.Length == 0 || action.Length == 0 ? "" : resource + ":" + action;
        }

        public static bool IsAuthorized(IEnumerable<Claim> claims, string resource, string action)
        {
            var list = claims?.ToList() ?? new List<Claim>();
            if (list.Any(c => c.Type == "org_role" && c.Value == "owner"))
            {
                return true;
            }
            var permission = PermissionFor(resource, action);
            return list.Any(c => c.Type == "org_permissions" && c.Value == permission);
        }

        public static Func<HttpContext, Func<Task>, Task> RequireAuthorization(string resource, string action)
        {
            return async (context, next) =>
            {
                if (!IsAuthorized(context.User?.Claims, resource, action))
                {
                    context.Response.StatusCode = StatusCodes.Status403Forbidden;
                    context.Response.ContentType = "application/json";
                    await context.Response.WriteAsync("{\"error\":\"forbidden\"}");
                    return;
                }
                await next();
            };
        }
    }
}
